### PR TITLE
Boolean && should either always break or never break

### PR DESCRIPTION
**Pull request by jmorrell**
First pass at implementing #89 

This isn't ready to be merged yet, and I'm probably going to rewrite all of it, but I'm putting it up for discussion in case anyone has any insight on the problems I'm seeing.

### CODE REVIEW

File: src/printer.js [hunk 1]
```diff
@@ -238,12 +238,65 @@ function genericPrintNoParens(path, options, print) {
     );
   case "BinaryExpression":
   case "LogicalExpression":
+    var parent = path.getParentNode();
+    var isStartOfLogicalExpressionChain = parent ? parent.type !== 'LogicalExpression' : false;
```
Comment (jmorrell) [inline src/printer.js]: We have to know when we're entering a LogicalExpression change vs recursing in the middle of a chain. This was my first idea, but it fails when you have a LogicalExpression in the middle of a LogicalExpression.

Ex:
```js
parent.type === "MemberExpression"
&& isNumber.check(node.value)
&& name === "object"
&& (
  nestedOne || nestedTwo || nestedThree || nestedFour || nestedFive || nestedSix || nestedSeven || nestedEight
);
```

I'd expect this to be formatted as:

```js
(
  parent.type === "MemberExpression" &&
  isNumber.check(node.value) &&
  name === "object" &&
  (
    nestedOne || 
    nestedTwo || 
    nestedThree || 
    nestedFour || 
    nestedFive || 
    nestedSix || 
    nestedSeven || 
    nestedEight
  )
)
```

I'm unsure how to detect the second group...

File: src/printer.js [hunk 2]
```diff
@@ -2051,3 +2104,21 @@ function isLastStatement(path) {
   const body = parent.body;
   return body && body[body.length - 1] === node;
 }
+
+function collectLogicalExpressions(node, input=[]) {
```
Comment (jmorrell) [inline src/printer.js]: There's probably a higher order function that could be formulated here, but I'll probably rewrite this following the template in `printMemberChain` which seems to face similar issues

File: tests/intersection/__snapshots__/jsfmt.spec.js.snap [hunk 3]
```diff
@@ -87,13 +87,19 @@ type DuplexStreamOptions =
...
+    options.objectMode ||
+    options.readableObjectMode ||
+    options.writableObjectMode
+  ); // error, undefined ~> boolean
 }
 
 function hasObjectMode_ok(options: DuplexStreamOptions): boolean {
-  return !!(options.objectMode || options.readableObjectMode ||
-    options.writableObjectMode);
+  return !!((
```
Comment (jmorrell) [inline tests/intersection/__snapshots__/jsfmt.spec.js.snap]: We need to know when to add parens and when not to 🤔

File: tests/prettier/__snapshots__/jsfmt.spec.js.snap [hunk 4]
```diff
@@ -1,3 +1,106 @@
...
+
+function b() {
+  if (isTrue() || isFalse()) {
+    return this;
+  }
+}
+
+function c() {
+  if (
+    (
```
Comment (vjeux) [inline tests/prettier/__snapshots__/jsfmt.spec.js.snap]: this parenthesis is no longer needed right?

File: tests/prettier/__snapshots__/jsfmt.spec.js.snap [hunk 5]
```diff
@@ -1,3 +1,106 @@
...
+  ) {
+    return this;
+  }
+}
+
+function d() {
+  switch (1) {
+    case \"Literal\":
+      return (
+        parent.type === \"MemberExpression\" &&
```
Comment (vjeux) [inline tests/prettier/__snapshots__/jsfmt.spec.js.snap]: good thing that it puts the && at the end :)

File: tests/prettier/__snapshots__/jsfmt.spec.js.snap [hunk 6]
```diff
@@ -1,3 +1,106 @@
...
+        isNumber.check(node.value) &&
+        name === \"object\" &&
+        nestedOne || nestedTwo || nestedThree || nestedFour || nestedFive || nestedSix || nestedSeven || nestedEight
```
Comment (vjeux) [inline tests/prettier/__snapshots__/jsfmt.spec.js.snap]: parenthesis is missing here

File: tests/prettier/__snapshots__/jsfmt.spec.js.snap (hunk 4 above)
Comment (jmorrell) [inline tests/prettier/__snapshots__/jsfmt.spec.js.snap]: Worse, I'm currently adding the paren :( Need to figure out under what cases I can omit it

File: tests/prettier/__snapshots__/jsfmt.spec.js.snap (hunk 6 above)
Comment (jmorrell) [inline tests/prettier/__snapshots__/jsfmt.spec.js.snap]: See my first code comment. I need to detect this as a second, nested expression somehow.

File: tests/intersection/__snapshots__/jsfmt.spec.js.snap (hunk 3 above)
Comment (jlongster) [inline tests/intersection/__snapshots__/jsfmt.spec.js.snap]: See `needsParens` in `fast-path.js`. That's where all the logic is for detecting if parens needs to be inserting or not for expressions.

Comment (jmorrell) [inline tests/intersection/__snapshots__/jsfmt.spec.js.snap]: I'll check it out 👍

**Comment by jlongster** [comment]
Thanks for the PR! You dug into a hairier problem and got pretty far, good job. I think we can make this code clearer though, as well as improve performance. Some of the problems I see:

* `collectLogicalExpressions` is recursive, and while I love recursion and functional programming, I'm very sensitive about perf in this project and it will easy degrade if we're not careful everywhere (death my a thousand cuts). We should do imperative iteration. Also the `pairs` machinery looks like a lot of work; I think we can simplify it.
* I'm not sure why we need `isStartOfLogicalExpressionChain`? Can't you just check if `chain.length > 1` ?
* Do we not want this for binary expressions as well? Maybe not, but we need to solve similar problems for binary expressions... we could start with just logical though.

I had already started this work on another branch. Here's my complete implementation so far:

```js
  case "BinaryExpression":
  case "LogicalExpression":
    const binaryExprs = []
    let curr = n;
    while(curr.type === "BinaryExpression") {
      binaryExprs.push({ right: curr.right, op: curr.operator });
      curr = curr.left;
    }
    binaryExprs.reverse();

    if(binaryExprs.length > 0) {
      return group(concat([
        // TODO: create correct paths
        print(FastPath.from(curr)),
        indent(options.tabWidth,
               concat(binaryExprs.map(expr => {
                 return concat([" ", expr.op, line, print(FastPath.from(expr.right))]);
               })))
      ]));
    }

    return group(concat([
        path.call(print, "left"),
        " ",
        n.operator,
        indent(options.tabWidth, concat([ line, path.call(print, "right") ]))
    ]));
```

While I need to fix a few things, and it applies it to binary expressions too, I believe this achieves similar results with a lot less code. Do you mind if I take this bug because there were a few things I wanted to look into while doing this?

**Comment by jmorrell** [comment]
Agreed on all points.

I have another version form this morning that gets rid of `isStartOfLogicalExpressionChain` and uses `multilineGroup` instead of `conditionalGroup`, and is way less code, but there are still some more issues with it that I haven't figured out yet. If you have something that's close I'm happy to let you take it :)

**Comment by jlongster** [comment]
One thing I wanted to figure out was getting rid of those `FastPath.from` calls because that breaks a few things because the nodes don't have context anymore (i.e. a link back to the parent). We are using that in 1 or 2 other places. I think there was a couple other things I wanted to look at too.

Hope this doesn't discourage you, there are several other bugs that would be great to have fixed!

**Comment by vjeux** [comment]
I can't wait to have this feature in!

**Pull request closed without merging**